design/object-bucket/object-bucket.md

```diff
@@ -95,8 +100,8 @@ _As a Kubernetes user, I want to leverage the Kubernetes API to create object st
 _As a Kubernetes user, I want to delete ObjectBucketClaim instances and cleanup generated API resources._
 
 1. The user deletes the OBC via `kubectl delete objectbucketclaim -n default FOO`.
-1. The OBC is marked for deletion and left in the foreground.
-1. The respective ConfigMap and user Secret are deleted.
+1. The OBC is marked for deletion and left in the foreground and the `Delete()` method, implemented by this provisioner, is invoked by the bucket lib.
```

Comment:
This would mean that the physical bucket is deleted.

Reply:
Agree. Should the bkt lib dictate the Delete semantics? I was thinking that the lib always calls `Delete()` and the provisioner decides if the physical bkt is deleted or not. Then the lib deletes the bkt related artifacts. The ceph-rgw provisioner can decide to do nothing when `Delete()` is called, thus preserving the physical bkt.

Reply:
A cursory at sig-storage-lib-external-provisioners gives me the impression that Delete events for PVCs and PVs are mostly ignored.

See the PVC [Delete handler](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go#L557) and the PV [Delete handler](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go#L584).  You'll see that the PVC handler is a noop on Deletes.  The PV handler's only operation is to instruct the queue to forget the key.  This means that the queue stops tracking the retries for the item before discarding it.

Now, that said, there are still large blocks of code responsible for actually deleting volumes base on a range of qualifiers (reclaimPolicy, whether it was dynamically provisioned, etc).  It may be that this is mostly dead code.  Jan just recently made the change to [ignore delete events](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/commit/c19e26be3fd97c9f9d97a47d5198fdfa14a38aa9#diff-18de8fd9ec37aa97da8831f9408ce6aeR560).

Orthogonally, since the Delete path is completely omitted, the controller performs no clean up tasks WRT API object.  Nor does it appear to perform status updates for relationship changes between the PV and PVC.  

cc @jsafrane - can you give us a true || false on the above?

Reply:
> A cursory at  gives me the impression that Delete events for PVCs and PVs are mostly ignored.

Yes, sig-storage-lib-external-provisioners ignores delete events, however, they're processed by Kubernetes (PV controller). If a PVC is deleted, PV is marked as Released and PV's reclaim policy is executed. If the policy is Delete, sig-storage-lib-external-provisioners gets an event that PV is Released and deletes the storage asset + the PV.

In addition, there is periodic sync in PV controller that checks that all PVs still have have corresponding PVCs. If a PVC is missing (i.e. the controller missed Delete event), again, the reclaim policy is executed.

Btw, you should not depend on Delete events from Kubernetes API at all. It's ok if you process them, but you won't get any Delete event when your operator / controller is not running or not accessible, e.g. the operator is being updated, network partition, the operator crashes etc. After such "hiccup", new operator is started and you should check that every bucket has its claim.

Reply:
@jsafrane thanks for your reply! If _reclaimPolicy_ == "delete" is it up to the external provisoiner to decide if the underlying physical storage is deleted? Does kubernetes have any influence on physical delete decisions?